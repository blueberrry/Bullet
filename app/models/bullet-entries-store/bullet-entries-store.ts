import { Instance, SnapshotIn, SnapshotOut, types } from "mobx-state-tree"
import "react-native-get-random-values"
import { v4 as uuidv4 } from "uuid"
import { BulletEntryApi } from "../../services/api/bullet-entry-api"
import { BulletEntry, BulletEntryModel } from "../bullet-entry/bullet-entry"
import { withEnvironment } from "../extensions/with-environment"
import { INITIAL_ALL_BULLET_ENTRIES } from "../initial-data/initial-data"
// import { INITIAL_ALL_BULLET_ENTRIES } from "../initial-data/initial-data"

/**
 * All bullet entries irrespective of date
 */
export const BulletEntriesStoreModel = types
  .model("BulletEntriesStore")
  .props({
    bulletEntries: types.optional(types.array(BulletEntryModel), []),
  })
  .extend(withEnvironment)
  .views((self) => ({
    // not allowed to change the model, only derive information from it
    // powered behind the scenes, tracks which data is relevant and only calculated when necessary
    get totalBulletEntries() {
      return self.bulletEntries.length
    },
  }))
  .actions((self) => ({
    saveBulletEntries: (bulletEntriesSnapshot: BulletEntry[]) => {
      self.bulletEntries.replace(bulletEntriesSnapshot)
    },
  }))
  .actions((self) => ({
    getBulletEntries: async () => {
      const bulletEntryApi = new BulletEntryApi(self.environment.api)
      const result = await bulletEntryApi.getBulletEntries()

      if (result.kind === "ok") {
        self.saveBulletEntries(result.bulletEntries)
      } else {
        __DEV__ && console.tron.log(result.kind)
      }
    },
  }))
  .actions((self) => ({
    // TODO: Causes tests to fail
    getInitialBulletEntriesForTesting: () => {
      // TODO: Temp starting point in lieue of api request
      if (self.bulletEntries.length < 3) {
        self.bulletEntries.replace(INITIAL_ALL_BULLET_ENTRIES.results)
      }
    },
  }))
  .actions((self) => ({
    addBulletEntry: ({ text = "", status = "todo" }) => {
      debugger
      const id = uuidv4()
      const newBulletEntry = BulletEntryModel.create({
        id,
        text,
        status,
        dateCreated: Date.now(),
      })
      self.bulletEntries.push(newBulletEntry)
    },
    // TODO: Unused
    getEntriesFromIds: (ids = []) => {
      // should be in view
      // pass array of guids for specific date
      // return relevant entries from all bullet entires
      const entriesFromIds = self.bulletEntries.filter((entry) => {
        if (ids.includes(entry.id)) {
          return entry
        }
        return null
      })

      return entriesFromIds
    },
  }))

export interface BulletEntriesStore extends Instance<typeof BulletEntriesStoreModel> {}
export interface BulletEntryStoreSnapshotOut extends SnapshotOut<typeof BulletEntriesStoreModel> {}
export interface BulletEntryStoreSnapshotIn extends SnapshotIn<typeof BulletEntriesStoreModel> {}
export const createBulletEntryStoreDefaultModel = () => types.optional(BulletEntriesStoreModel, {})
