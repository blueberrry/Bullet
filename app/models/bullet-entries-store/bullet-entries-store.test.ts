import { reaction } from "mobx"
import { getSnapshot, onPatch, onSnapshot } from "mobx-state-tree"
import { BulletEntriesStoreModel } from "./bullet-entries-store"

it("can create a bullet entries list", () => {
  const bulletEntriesList = BulletEntriesStoreModel.create({
    bulletEntries: [
      {
        id: "unique-id-string",
        status: "todo",
        text: "Bullet entry text",
        dateCreated: 1662047699328,
      },
    ],
  })

  expect(bulletEntriesList.bulletEntries.length).toBe(1)
  expect(bulletEntriesList.bulletEntries[0].id).toBe("unique-id-string")
  expect(bulletEntriesList.bulletEntries[0].status).toBe("todo")
  expect(bulletEntriesList.bulletEntries[0].text).toBe("Bullet entry text")
  expect(bulletEntriesList.bulletEntries[0].dateCreated).toBe(1662047699328)
})

it("can add new items", () => {
  const bulletEntriesList = BulletEntriesStoreModel.create()
  // Already defined in data model so we don't need to do BulletEntry.create()
  // mobxstatetree will create the model for us
  // Pass plain JSON invoke actions on the model instance

  const states = []

  onSnapshot(bulletEntriesList, (snapshot) => {
    states.push(snapshot)
  })

  bulletEntriesList.addBulletEntry({
    id: "unique-id-string-2",
    status: "inspirationalIdeas",
    text: "Build a plant wall",
    dateCreated: 123456,
  })

  expect(bulletEntriesList.bulletEntries.length).toBe(1)
  expect(bulletEntriesList.bulletEntries[0].id).toBe("unique-id-string-2")
  expect(bulletEntriesList.bulletEntries[0].status).toBe("inspirationalIdeas")
  expect(bulletEntriesList.bulletEntries[0].text).toBe("Build a plant wall")

  bulletEntriesList.bulletEntries[0].changeStatus("note")
  expect(bulletEntriesList.bulletEntries[0].status).toBe("note")

  // getSnapeshot is the inverserse process of what create does

  // expect(getSnapshot(bulletEntriesList)).toEqual({
  //   bulletEntries: [
  //     {
  //       id: "unique-id-string-2",
  //       status: "note",
  //       text: "Build a plant wall",
  //       dateCreated: 123456,
  //     },
  //   ],
  // })

  expect(getSnapshot(bulletEntriesList)).toMatchSnapshot()

  expect(states).toMatchSnapshot()
})

it("can change the text", () => {
  const bulletEntriesList = BulletEntriesStoreModel.create()
  const patches = [] // when tree changes, what a mutation looks like
  // very powerful for communication with backend
  onPatch(bulletEntriesList, (patch) => {
    patches.push(patch)
  })

  bulletEntriesList.addBulletEntry({
    id: "unique-id-string-3",
    status: "inspirationalIdeas",
    text: "Build a plant wall",
    dateCreated: 123456,
  })

  bulletEntriesList.bulletEntries[0].changeText("Visit low light polution area")

  expect(patches).toMatchSnapshot()
})

it("can change the status", () => {
  const bulletEntriesList = BulletEntriesStoreModel.create()
  const patches = [] // when tree changes, what a mutation looks like
  // very powerful for communication with backend
  onPatch(bulletEntriesList, (patch) => {
    patches.push(patch)
  })

  bulletEntriesList.addBulletEntry({
    id: "unique-id-string-3",
    status: "inspirationalIdeas",
    text: "Build a plant wall",
    dateCreated: 123456,
  })

  bulletEntriesList.bulletEntries[0].changeStatus("todo")

  expect(patches).toMatchSnapshot()
})

it("can calculate the total bullet entries", () => {
  const bulletEntriesList = BulletEntriesStoreModel.create({
    bulletEntries: [
      {
        id: "unique-id-string-3",
        status: "todo",
        text: "Mow the lawn",
        dateCreated: 123456,
      },
      {
        id: "unique-id-string-4",
        status: "todo",
        text: "Play tenis",
        dateCreated: 123456,
      },
    ],
  })

  expect(bulletEntriesList.totalBulletEntries).toBe(2)

  let changed = 0

  reaction(
    () => bulletEntriesList.bulletEntries.length,
    () => changed++,
  )

  const snapshots = []

  expect(changed).toBe(0)

  onSnapshot(bulletEntriesList, (snapshot) => {
    snapshots.push(snapshot)
  })

  bulletEntriesList.addBulletEntry({
    id: "unique-id-string-5",
    status: "todo",
    text: "Research Jarrariums",
    dateCreated: 123456,
  })

  expect(changed).toBe(1)
})

// TODO: Stub date.now() global https://codewithhugo.com/mocking-the-current-date-in-jest-tests/
