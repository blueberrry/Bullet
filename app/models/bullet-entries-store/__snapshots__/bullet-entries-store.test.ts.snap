// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items 1`] = `
Object {
  "bulletEntries": Array [
    Object {
      "dateCreated": 123456,
      "id": "unique-id-string-2",
      "status": "note",
      "text": "Build a plant wall",
    },
  ],
}
`;

exports[`can add new items 2`] = `
Array [
  Object {
    "bulletEntries": Array [
      Object {
        "dateCreated": 123456,
        "id": "unique-id-string-2",
        "status": "inspirationalIdeas",
        "text": "Build a plant wall",
      },
    ],
  },
  Object {
    "bulletEntries": Array [
      Object {
        "dateCreated": 123456,
        "id": "unique-id-string-2",
        "status": "note",
        "text": "Build a plant wall",
      },
    ],
  },
]
`;

exports[`can change the status 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/bulletEntries/0",
    "value": Object {
      "dateCreated": 123456,
      "id": "unique-id-string-3",
      "status": "inspirationalIdeas",
      "text": "Build a plant wall",
    },
  },
  Object {
    "op": "replace",
    "path": "/bulletEntries/0/status",
    "value": "todo",
  },
]
`;

exports[`can change the text 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/bulletEntries/0",
    "value": Object {
      "dateCreated": 123456,
      "id": "unique-id-string-3",
      "status": "inspirationalIdeas",
      "text": "Build a plant wall",
    },
  },
  Object {
    "op": "replace",
    "path": "/bulletEntries/0/text",
    "value": "Visit low light polution area",
  },
]
`;
